pageserver/src/lib.rs

```diff
@@ -85,6 +87,50 @@ pub async fn shutdown_pageserver(
     exit_code: i32,
 ) {
     use std::time::Duration;
+
+    let walredo_extraordinary_shutdown_thread_span = {
+        let span = info_span!(parent: None, "walredo_extraordinary_shutdown_thread");
+        span.follows_from(tracing::Span::current());
+        span
+    };
+    let walredo_extraordinary_shutdown_thread = std::thread::spawn(move || {
+        let _entered = walredo_extraordinary_shutdown_thread_span.enter();
+        std::thread::sleep(Duration::from_secs(8));
+        let managers = tenant::WALREDO_MANAGERS
+            .lock()
+            .unwrap()
+            // prevents new walredo managers from being inserted
+            .take()
+            .expect("only we take()");
+        // Use FuturesUnordered to get in queue early for each manager's
+        // heavier_once_cell semaphore wait list.
+        // Also, for idle tenants that for some reason haven't
+        // shut down yet, it's quite likely that we're not going
+        // to get Poll::Pending once.
+        let mut futs: FuturesUnordered<_> = managers
+            .into_iter()
+            .filter_map(|(_, mgr)| mgr.upgrade())
+            .map(|mgr| async move { tokio::task::unconstrained(mgr.shutdown()).await })
+            .collect();
+        if futs.is_empty() {
+            info!("no walredo managers left");
+            return;
+        }
+        info!(count=%futs.len(), "starting shutdown");
```

Comment:
it would be nicer for aggregation tooling I think if this were before the if (maybe with a at that point slightly more fitting message like "built walredo list").

Reply:
623beb0ef96d8c23291f556d72b6d19006f3c6cf